### ISSUE TITLE
[Feature] Audio Stories

### ISSUE THREAD

**Issue by binarygeek119** [stashapp/stash#1258]
there is a issue for this but as im putting a bounty on it thought i should make mine new.
old issue #11
how i see this working is,
is to play a mp3,ogg,wav,m4a,etc in the Brower
metadata i like to see is 
name
author
Performer
url for source
website of source
date
language 
cover front
cover back
Rating
Tags
Details

**Comment by elad-lachmi** [comment]
I'm willing to give this one a _go_ (see what I did there... meh).
Anyway, might take me a little bit of time, as I haven't done anything significant in go for a while.
If that's all good, I'll take it.

**Comment by WithoutPants** [comment]
$50 bounty added (txn [#793013](https://opencollective.com/stashapp/contributions/739013)). Total bounty $90.

**Comment by miltuss** [comment]
Please. Publish this update which is of paramount importance to stash!

**Comment by styxmaster06** [comment]
I'd also love this. I have a huge audio library that I'd like to use here, and I have a couple additional thoughts on how it should be implemented.

**_Multiple Source/website URL entries._** GonewildAudio, for example, has a reddit link and potentially several other sources. It would be good to be able to add all the links in a post, possibly by scraper eventually, so that if you need to go back and re-source the file or something, you can get it even if one of the sources is down or hidden.)

**_Grouping Audios._** Sometimes audios come as parts of a whole, so it would be nice to tag them as such, similar to how Movies are groups of Scenes. I would propose calling it "Series", but the name doesn't matter much.

**Comment by Hydronicx** [comment]
This would be extremely helpful in organizing audio collections from r/GoneWildAudio.
Why not implement the simplest support for audio files to be played as Audios? The rest would be identical to scenes with some minor differences.

**Comment by d0t-d0t-d0t** [comment]
If anyone is interested I've made a temporary solution:

1. Add audio extensions in the library settings.
2. Create an "Audio" tag.
3. Make and "Audio" directory to put the files in.
4. Use Auto Tag to tag the files inside the folder.
5. Use custom css and js to modify or replace the player (like below).

Here are my custom css and js:
css:
```css
/* Audio: Hide player. */
.VideoPlayer.audio .vjs-big-play-button, .VideoPlayer.audio .scrubber-wrapper, .VideoPlayer.audio .vjs-fullscreen-control,
.VideoPlayer.audio .vjs-control-bar {
  display: none !important;
}
```

js:
```js

// Replace player
function setAudioElement(bool){
  const pl = document.querySelector(".VideoPlayer")
  var elm = document.getElementById('VideoJsPlayer_html5_api');
  if (bool){
    const videoElement = elm
    pl.classList.add("audio")
    var audioElement = document.createElement('audio');
    for (var i = 0; i < videoElement.attributes.length; i++) {
        audioElement.setAttribute(videoElement.attributes[i].name, videoElement.attributes[i].value);
    }
    audioElement.src = videoElement.src
    audioElement.controls = true
    audioElement.autoplay = false
    videoElement.parentNode.replaceChild(audioElement, videoElement);
    audioElement.parentNode.insertBefore(document.querySelector(".vjs-poster"), audioElement);
  }else{
    const audioElement = elm
    pl.classList.remove("audio")
    var videoElement = document.createElement('video');
    for (var i = 0; i < audioElement.attributes.length; i++) {
      videoElement.setAttribute(audioElement.attributes[i].name, audioElement.attributes[i].value);
    }
    videoElement.src = audioElement.src
    videoElement.controls = false
    audioElement.parentNode.replaceChild(videoElement, audioElement);
    videoElement.parentNode.insertBefore(videoElement, document.querySelector(".vjs-poster"));
  }
}


// Add "Audio Only" switch button
function audioswitch(){
  const elm = document.querySelector(".scene-toolbar-group")
  const btng = document.createElement("div")
  const btn = document.createElement("button")
  const pl = document.querySelector(".VideoPlayer")
  if (pl.classList.contains("audio")) {
    btn.style.background = "rgba(138,155,168,.15)"
  }
  btn.classList.add("minimal", "btn", "btn-secondary")
  btn.innerHTML= "Audio Only"
  btn.id = "audioSwitch"
  btn.onclick = function (){
    const pl = document.querySelector(".VideoPlayer")
    const bt = document.getElementById("audioSwitch")
    if (pl.classList.contains("audio")){
      setAudioElement(false);
      bt.style.background = "transparent"
    } else {
      setAudioElement(true);
      bt.style.background = "rgba(138,155,168,.15)"
    }
  }
  btng.classList.add("btn-group")
  btng.appendChild(btn)
  elm.appendChild(btng)
}


// Find audio tag and replace player
function findAudioTag() {
  const links = document.querySelectorAll('.tag-item a');
if(links){
  links.forEach(link => {
      const divText = link.querySelector('div').textContent.trim();

      if (divText === "Audio") {
        setAudioElement(true)
      }
  });}
}

// Check if page is a scene page
function shouldRunOnPage() {
    const currentPath = window.location.pathname;
    const itemsPagePattern = /^\/scenes\/\d+(.*)/;
    return itemsPagePattern.test(currentPath);
}

// Check if elements are rendered every 100ms then
function checkRendered() {
  const targetElement2 = document.querySelector('.VideoPlayer');
  const targetElement3 = document.querySelector('#VideoJsPlayer_html5_api')?.src
  const targetElement4 = document.querySelector('.scene-toolbar-group')
  if (targetElement2 && targetElement3() && targetElement4) {
    findAudioTag()
    audioswitch()
    clearInterval(intervalId);
  }
}

const intervalId = setInterval(checkRendered, 100);
```

**Comment by miltuss** [comment]
> Here are my custom css and js:

@d0t-d0t-d0t I don't know what this code is supposed to do, but it doesn't work with stash version 0.27. No video/audio player replacement has been observed on audio files with the `Audio` tag. Can you help resolve this issue?

**Comment by d0t-d0t-d0t** [comment]
> > Here are my custom css and js:
> 
> @d0t-d0t-d0t I don't know what this code is supposed to do, but it doesn't work with stash version 0.27. No video/audio player replacement has been observed on audio files with the `Audio` tag. Can you help resolve this issue?

You should try refreshing the page, the code isn't perfect.

This is the updated code:
```js
function setAudioElement(bool){
  const pl = document.querySelector(".VideoPlayer")
  var elm = document.getElementById('VideoJsPlayer_html5_api');
  if (bool){
    const videoElement = elm
    pl.classList.add("audio")
    var audioElement = document.createElement('audio');
    for (var i = 0; i < videoElement.attributes.length; i++) {
        audioElement.setAttribute(videoElement.attributes[i].name, videoElement.attributes[i].value);
    }
    audioElement.src = videoElement.src.replace(/(^.*\/scene\/\d+\/)(stream.*)\?/, "$1stream?")
    audioElement.controls = true
    audioElement.autoplay = false
    videoElement.parentNode.replaceChild(audioElement, videoElement);
    audioElement.parentNode.insertBefore(document.querySelector(".vjs-poster"), audioElement);
  }else{
    const audioElement = elm
    pl.classList.remove("audio")
    var videoElement = document.createElement('video');
    for (var i = 0; i < audioElement.attributes.length; i++) {
      videoElement.setAttribute(audioElement.attributes[i].name, audioElement.attributes[i].value);
    }
    videoElement.src = audioElement.src
    videoElement.controls = false
    audioElement.parentNode.replaceChild(videoElement, audioElement);
    videoElement.parentNode.insertBefore(videoElement, document.querySelector(".vjs-poster"));
  }
}



function audioswitch(){
  const elm = document.querySelector(".scene-toolbar-group")
  const btng = document.createElement("div")
  const btn = document.createElement("button")
  const pl = document.querySelector(".VideoPlayer")
  if (pl.classList.contains("audio")) {
    btn.style.background = "rgba(138,155,168,.15)"
  }
  btn.classList.add("minimal", "btn", "btn-secondary")
  btn.innerHTML= "Audio Only"
  btn.id = "audioSwitch"
  btn.onclick = function (){
    const pl = document.querySelector(".VideoPlayer")
    const bt = document.getElementById("audioSwitch")
    if (pl.classList.contains("audio")){
      setAudioElement(false);
      bt.style.background = "transparent"
    } else {
      setAudioElement(true);
      bt.style.background = "rgba(138,155,168,.15)"
    }
  }
  btng.classList.add("btn-group")
  btng.appendChild(btn)
  elm.appendChild(btng)
}



function findAudioTag() {
  // Select all <a> elements within the <span>
  const links = document.querySelectorAll('.tag-item a');
if(links){
  // Iterate through each <a> element
  links.forEach(link => {
      // Get the text content of the <div> inside the <a>
      const divText = link.querySelector('div').textContent.trim();

      // Check if the text matches the target text
      if (divText === "Audio") {
        setAudioElement(true)
      }
  });}
}

function shouldRunOnPage() {
    const currentPath = window.location.pathname;
    const itemsPagePattern = /^\/scenes\/\d+(.*)/;
    return itemsPagePattern.test(currentPath);
}

if (shouldRunOnPage()){
function checkRendered() {
  const targetElement2 = document.querySelector('.VideoPlayer');
  const targetElement3 = document.querySelector('#VideoJsPlayer_html5_api')?.src
  const targetElement4 = document.querySelector('.scene-toolbar-group')
  if (targetElement2 && targetElement3 && targetElement4) {
    findAudioTag()
    audioswitch()
    // Stop polling
    clearInterval(intervalId);
  }
}


    const intervalId = setInterval(checkRendered, 100);
}
```

Remember to use this custom css:
```css
/* Audio: Remove big play button (leave only the button in controls). */
.VideoPlayer.audio .vjs-big-play-button, .VideoPlayer.audio .scrubber-wrapper, .VideoPlayer.audio .vjs-fullscreen-control,
.VideoPlayer.audio .vjs-control-bar {
  display: none !important;
}

/* Audio: Make the controlbar visible by default */
.VideoPlayer.audio .vjs-control-bar {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}
/* Make player height minimum to the controls height so when we hide video/poster area the controls are displayed correctly. */
/*
.VideoPlayer.audio .video-js {
  height: fill !important;
}

.VideoPlayer.audio .video-wrapper {
  height: 6rem !important;
}
*/
```

**Comment by miltuss** [comment]
@d0t-d0t-d0t I updated the code, and refreshed the page several times, I even cleared the browser cache and restarted the docker container several times. CCSS and custom javascript are activated in the settings but the code changes absolutely nothing to the reader!

**Comment by d0t-d0t-d0t** [comment]
@miltuss  You should see something like this:
![image-1](https://github.com/user-attachments/assets/74913f1c-c7bf-4544-aa4a-f356d9a42b46)

There should also be an 'Audio Only' toggle button

**Comment by miltuss** [comment]
@d0t-d0t-d0t Unfortunately your script only works for you

**Comment by d0t-d0t-d0t** [comment]
That's strange,  you could try to copy the script and paste it in the browser console to see if it works, i was thinking of making a plugin (i don't know if plugins can use js to modify preexisting elements) if i had the time or maybe a simpler user script.

**Comment by miltuss** [comment]
@d0t-d0t-d0t I figured out what's wrong with the code, when we navigate to stash to open an audio file with the Audio Tag, the script doesn't run like this:

![Sélection_560](https://github.com/user-attachments/assets/40a08752-93db-40ee-b098-62f7ce74a780)

The script necessarily needs a refresh of the reader page to run.

![Sélection_561](https://github.com/user-attachments/assets/e0dfed68-90ed-4f82-b092-ba1060381c58)

In addition, markers do not work with the browser's built-in reader. So the method consisting of using the Stash player with the HLS transcoder remains the best while waiting for real integration of audio support in Stash.

**Comment by d0t-d0t-d0t** [comment]
@miltuss 
>I figured out what's wrong with the code, when we navigate to stash to open an audio file with the Audio Tag, the script doesn't run like this:
I forgot to mention that you have to replace the match with the url of your stash instance.

>In addition, markers do not work with the browser's built-in reader.
The code can be modified to use the stash player but i haven't tested it with audio files.

**Comment by miltuss** [comment]
@d0t-d0t-d0t 

> I forgot to mention that you have to replace the match with the url of your stash instance.

I had inserted the URL of my instance instead of the default URL of the userscript from the first test

> The code can be modified to use the stash player but i haven't tested it with audio files.

The player must read the audio correctly provided the HLS codec is selected. I don't know if it is possible to code a script that can automatically select the HSL on stash's Video.JS player when playing an audio file. I think this would be the best DIY to do while waiting for the official release of this feature!

**Comment by Tweeticoats** [comment]
I have created the audio-transcodes plugin for stash.
This plugin converts an audio file to a video and stash will play this video instead allowing the normal video player to work.
Install the plugin and run tasks > audio-transcodes > Process All

**Comment by miltuss** [comment]
By doing this, your audio files will not be recognized by stash as audio tracks when integrating the AudioStoris functionality into Stash. 

HLS allows you to play audio files in Stash, you just need to find a way to exclude other formats by default when playing an audio file, so that playback starts immediately on HLS and not manually switch to HLS.

**Comment by d0t-d0t-d0t** [comment]
Here are the updated script/css and userscript that use the original player and automatically change the transcoder to his:

## Costum JS & CSS
```css
/* Stash Audio Player */
.VideoPlayer.audio .scrubber-wrapper {
    display: none !important;
}
.VideoPlayer.audio .vjs-control-bar {
    display: flex !important;
    opacity: 100 !important;
}
.VideoPlayer.audio .video-js {
    background-color: #202b33 !important;
}
    .VideoPlayer.audio .video-wrapper {
    height: 73.25vw !important;
}
.VideoPlayer.audio #VideoJsPlayer_html5_api,
#VideoJsPlayer > div.vjs-poster {
    height: 56.25vw !important;
}
```
```js
// Stash Audio Player
(function() {
    'use strict';
    function waitForElementClass(elementId, callBack, varPass, time) {
            time = (typeof time !== 'undefined') ? time : 100;
            window.setTimeout(() => {
                const element = document.getElementsByClassName(elementId);
                if (element.length > 0) {
                    callBack(varPass, elementId, element);
                } else {
                    waitForElementClass(elementId, callBack);
                }
            }, time);
        }

    function setAudioElement(bool) {
        const pl = document.querySelector(".VideoPlayer");
        var elm = document.getElementById('VideoJsPlayer_html5_api');
        const sourceElements = Array.from(document.querySelectorAll("#VideoJsPlayer > div.vjs-control-bar > div.vjs-source-selector.vjs-menu-button.vjs-menu-button-popup.vjs-control.vjs-button > div > ul > li > span.vjs-menu-item-text"));
        const hlsButton = sourceElements.find(el => el.textContent.trim().toLowerCase() === "hls");
        if (bool) {
            const videoElement = elm;
            pl.classList.add("audio");
            if (hlsButton){
                hlsButton.click()
            }
        } else {
            const audioElement = elm;
            pl.classList.remove("audio")
            sourceElements[0].click()
        }
    }

    function audioswitch() {
        const elm = document.querySelector(".scene-toolbar-group");
        const btng = document.createElement("div");
        const btn = document.createElement("button");
        const pl = document.querySelector(".VideoPlayer");
        if (pl.classList.contains("audio")) {
            btn.style.background = "rgba(138,155,168,.15)";
        }
        btn.id = "audioSwitch";
        btn.classList.add("minimal", "btn", "btn-secondary");
        btn.innerHTML = "Audio Only";
        btn.onclick = function() {
            const pl = document.querySelector(".VideoPlayer");
            const bt = document.getElementById("audioSwitch");
            if (pl.classList.contains("audio")) {
                setAudioElement(false);
                bt.style.background = "transparent";
            } else {
                setAudioElement(true);
                bt.style.background = "rgba(138,155,168,.15)";
            }
        };
        btng.classList.add("btn-group");
        btng.appendChild(btn);
        elm.appendChild(btng);
    }

    function findAudioTag() {
        const links = document.querySelectorAll('.tag-item a');
        if (links) {
            links.forEach(link => {
                const divText = link.querySelector('div').textContent.trim();
                if (divText === "Audio") {
                    setAudioElement(true);
                }
            });
        }
    }

    waitForElementClass("VideoPlayer", function() {
        waitForElementClass("tag-item ", findAudioTag);
        waitForElementClass("scene-toolbar-group", audioswitch);
    })

})();
```

## Userscript
```js
// ==UserScript==
// @name         Stash Audio Switch
// @version      0.2
// @description  Add an audio-only switch to video player and apply custom CSS
// @match        http://example:0000/scenes/*
// @grant        GM_addStyle
// ==/UserScript==

(function() {
    'use strict';
    function waitForElementClass(elementId, callBack, varPass, time) {
            time = (typeof time !== 'undefined') ? time : 100;
            window.setTimeout(() => {
                const element = document.getElementsByClassName(elementId);
                if (element.length > 0) {
                    callBack(varPass, elementId, element);
                } else {
                    waitForElementClass(elementId, callBack);
                }
            }, time);
        }

    GM_addStyle(`
        .VideoPlayer.audio .scrubber-wrapper {
            display: none !important;
        }
        .VideoPlayer.audio .vjs-control-bar {
            display: flex !important;
            opacity: 100 !important;
        }
        .VideoPlayer.audio .video-js {
            background-color: #202b33 !important;
        }
         .VideoPlayer.audio .video-wrapper {
            height: 73.25vw !important;
        }
        .VideoPlayer.audio #VideoJsPlayer_html5_api,
        #VideoJsPlayer > div.vjs-poster {
            height: 56.25vw !important;
        }
    `);

    function setAudioElement(bool) {
        const pl = document.querySelector(".VideoPlayer");
        var elm = document.getElementById('VideoJsPlayer_html5_api');
        const sourceElements = Array.from(document.querySelectorAll("#VideoJsPlayer > div.vjs-control-bar > div.vjs-source-selector.vjs-menu-button.vjs-menu-button-popup.vjs-control.vjs-button > div > ul > li > span.vjs-menu-item-text"));
        const hlsButton = sourceElements.find(el => el.textContent.trim().toLowerCase() === "hls");
        if (bool) {
            const videoElement = elm;
            pl.classList.add("audio");
            if (hlsButton){
                hlsButton.click()
            }
        } else {
            const audioElement = elm;
            pl.classList.remove("audio")
            sourceElements[0].click()
        }
    }

    function audioswitch() {
        const elm = document.querySelector(".scene-toolbar-group");
        const btng = document.createElement("div");
        const btn = document.createElement("button");
        const pl = document.querySelector(".VideoPlayer");
        if (pl.classList.contains("audio")) {
            btn.style.background = "rgba(138,155,168,.15)";
        }
        btn.id = "audioSwitch";
        btn.classList.add("minimal", "btn", "btn-secondary");
        btn.innerHTML = "Audio Only";
        btn.onclick = function() {
            const pl = document.querySelector(".VideoPlayer");
            const bt = document.getElementById("audioSwitch");
            if (pl.classList.contains("audio")) {
                setAudioElement(false);
                bt.style.background = "transparent";
            } else {
                setAudioElement(true);
                bt.style.background = "rgba(138,155,168,.15)";
            }
        };
        btng.classList.add("btn-group");
        btng.appendChild(btn);
        elm.appendChild(btng);
    }

    function findAudioTag() {
        const links = document.querySelectorAll('.tag-item a');
        if (links) {
            links.forEach(link => {
                const divText = link.querySelector('div').textContent.trim();
                if (divText === "Audio") {
                    setAudioElement(true);
                }
            });
        }
    }

    waitForElementClass("VideoPlayer", function() {
        waitForElementClass("tag-item ", findAudioTag);
        waitForElementClass("scene-toolbar-group", audioswitch);
    })

})();
```

**Comment by feederbox826** [comment]
This would be better suited for a [gist](https://gist.github.com/). Issue comments are not a good place for versioned code and take up an immense amount of vertical space

**Comment by d0t-d0t-d0t** [comment]
> This would be better suited for a [gist](https://gist.github.com/). 

I didn't know about gits, I've updated the comment.

**Comment by d0t-d0t-d0t** [comment]
I've made it in to a plugin:
https://github.com/d0t-d0t-d0t/StashAudioPlayer

**Comment by miltuss** [comment]
> I've made it in to a plugin: https://github.com/d0t-d0t-d0t/StashAudioPlayer

Hello, how do I add the plugin repository to Stash settings?
Many here use Docker and don't add plugins manually to the filesystem.

**Comment by d0t-d0t-d0t** [comment]
> Hello, how do I add the plugin repository to Stash settings? Many here use Docker and don't add plugins manually to the filesystem.

I've added instructions to the readme.

**Comment by d0t-d0t-d0t** [comment]
Now the audio player plugin is in my stash repo:
https://github.com/d0t-d0t-d0t/stash-repo

**Comment by rezreal** [comment]
There is an older attempt to support at least mp3s that, beside broken cover extraction, at least somehow worked for me:
https://github.com/stashapp/stash/commit/c4ed3ffd551bbc3ab004f5f72f2d27d5b84a4e9c

**Comment by RampantDespair** [comment]
Is this bounty still active or was it collected by the plugin developer?

**Comment by DogmaDragon** [comment]
> Is this bounty still active or was it collected by the plugin developer?

As long as issue is open the bounty is active.

**Comment by RampantDespair** [comment]
@DogmaDragon 

Well I got a few questions, just so I understand the actual request;
Is it that you guys wish to listen to porn video's audio track only?

Or do you want a fully fledged music player within the website?
Something like this?
![Image](https://github.com/user-attachments/assets/64e728fa-1dc6-4fd9-b517-f581e44863a1)
Because if that's the case, isn't that wayyyy out of scope for this project?
I could probably do it regardless, but that such a big leap that it would probably end up being it's own app?

**Comment by Hydronicx** [comment]
> [@DogmaDragon](https://github.com/DogmaDragon)
> 
> Well I got a few questions, just so I understand the actual request; Is it that you guys wish to listen to porn video's audio track only?
> 
> Or do you want a fully fledged music player within the website? Something like this? ![Image](https://github.com/user-attachments/assets/64e728fa-1dc6-4fd9-b517-f581e44863a1) Because if that's the case, isn't that wayyyy out of scope for this project? I could probably do it regardless, but that such a big leap that it would probably end up being it's own app?

No, not at all. This request is to be able to support pr0n that is audio-only. This is mostly for the amateur GoneWildAudio community on Reddit where people record adult audio files.

It would be nice to be able to keep all the functionalities of stash (performers, tags, metadata, etc.) and simply add the possibility of scanning and playing audio files (mp3, m4a, etc.). I don't think anybody actually cares to have a real "music" player-type thing. Just to be able to scan an audio library, play the files as "scenes" (some change in terminology here will probably be better) and then add metadata, scrape Reddit and soundgasm, etc.

**Comment by RampantDespair** [comment]
@Hydronicx 
Understood thank you for clarifying, I'll take a look at stash's stack and see what I can do.

In any case I suppose we would need to come up with a word that categorizes this type of content, because it will need a tab of its own from what I understand?

**Comment by Hydronicx** [comment]
> [@Hydronicx](https://github.com/Hydronicx) Understood thank you for clarifying, I'll take a look at stash's stack and see what I can do.
> 
> In any case I suppose we would need to come up with a word that categorizes this type of content, because it will need a tab of its own from what I understand?

I think "Audio" would work perfectly fine as a word. Having a tab for "Audios" would be great.

**Comment by DogmaDragon** [comment]
@RampantDespair 

> I'm thinking a new top-level object `audio` with a field list similar to video/images:
> 
> ```
> title
> url
> checksum
> path
> date
> details
> rating
> organized
> studio
> tags
> performers
> o_counter
> file {
>  size
>  duration
>  bitrate
> }
> ```
> 
> 
> 
> The library configuration would need to be adjusted to allow inclusion/exclusion of audio files. The current tick box way of doing things will likely be insufficient, so we'll need another approach.
> 
> 
> 
> We'll need `audioExtensions` and `excludedAudioPatterns` configuration fields.

 _Originally posted by @WithoutPants in [#11](https://github.com/stashapp/stash/issues/11#issuecomment-809915975)_

**Comment by edvinPL** [comment]
Interesting how one can tweak a PORN app into a really good Digital Asset Management software due to really good mouse-over playback. If only it supported audio with soundwaves = it would be a full-stack app for Video Editors

Perhaps even Porn Video Editors LOL
Studios could be clients/youtube channels

Only that audio thing...

**Comment by redwing509** [comment]
/attempt #1258

Implementation plan:
1. Analyze the existing codebase and understand the requirements
2. Implement the requested changes with clean, tested code
3. Add comprehensive tests covering the acceptance criteria
4. Submit PR with documentation updates if needed

Estimated delivery: within 24 hours.

**Comment by edvinPL** [comment]
> $50 bounty added (txn [#793013](https://opencollective.com/stashapp/contributions/739013)). Total bounty $90.

<img width="483" height="742" alt="Image" src="https://github.com/user-attachments/assets/5a1c1af6-2237-4b0c-abd1-ac95c07c1b9f" />
can I ask to add this to bounty as well?
Contribution ID #937599

**Comment by edvinPL** [comment]
<img width="1883" height="898" alt="Image" src="https://github.com/user-attachments/assets/db3bdf9d-e2c6-4515-8879-dd09d4332816" />
This preview scrubbing is literally the very best Media Asset Manager I have found over the internet.
IF only it had audio preview/scrubbing/playback with sound and soundwaves - this would be all I'd need :) 💯 🥇

**Comment by allornothingai** [comment]
To address the feature request for audio file/story support (Issue #1258), a new `Audio` media type needs to be introduced into the Stash application. This involves defining a new GraphQL schema, implementing corresponding backend logic for file scanning, metadata extraction, database storage, and developing a frontend UI for managing and playing audio files.

Below is a proposal for the core GraphQL schema changes and an example of how plugins would interact with the new API.

### 1. Proposed GraphQL Schema (`audio.graphql`)

This new file would be added under `graphql/schema/` (e.g., `graphql/schema/audio.graphql`). This schema defines the `Audio` type and its associated input types, queries, and mutations, incorporating all the requested metadata fields.

```graphql
# Defines the Audio object
type Audio {
  id: ID!
  checksum: String! # Unique identifier for the audio file content (e.g., SHA256)
  path: String! # Full path to the audio file
  size: Int! # File size in bytes
  duration: Float # Duration of the audio in seconds

  created: Timestamp!
  updated: Timestamp!

  # User-editable metadata
  name: String
  author: String # Primary author or narrator, could be a string
  performers: [Performer!] # Array of associated performers (many-to-many relationship)
  url: String # URL to the source of the audio
  website: String # Website of the source
  date: String # Date of creation or release (e.g., YYYY-MM-DD)
  language: String # Language of the audio (e.g., "en", "es")

  cover_front: Image # Front cover image (linked via Image ID)
  cover_back: Image # Back cover image (linked via Image ID)

  rating: Int # User rating
  tags: [Tag!] # Array of associated tags (many-to-many relationship)
  details: String # Detailed description or notes
}

# Input for creating a new Audio object
input AudioCreateInput {
  path: String!
  checksum: String!
  size: Int!
  duration: Float # Optional
  
  name: String
  author: String
  performerIds: [ID!] # IDs of existing performers
  url: String
  website: String
  date: String
  language: String
  coverFrontId: ID # ID of an existing Image for the front cover
  coverBackId: ID # ID of an existing Image for the back cover
  rating: Int
  tagIds: [ID!] # IDs of existing tags
  details: String
}

# Input for updating an existing Audio object
input AudioUpdateInput {
  id: ID!
  
  name: String
  author: String
  performerIds: [ID!]
  url: String
  website: String
  date: String
  language: String
  coverFrontId: ID
  coverBackId: ID
  rating: Int
  tagIds: [ID!]
  details: String
}

# Input for filtering audio objects
input AudioFilterType {
  q: String # General search query (name, author, details, etc.)
  tags: IDFilterType # Filter by associated tags
  performers: IDFilterType # Filter by associated performers
  # Additional filters can be added here (e.g., duration range, language filter)
}

# Extend the Query type to include audio-related queries
extend type Query {
  findAudios(filter: AudioFilterType, first: Int, skip: Int, sort: FindSortType): [Audio!]!
  getAudio(id: ID!): Audio
}

# Extend the Mutation type to include audio-related mutations
extend type Mutation {
  audioCreate(input: AudioCreateInput!): Audio!
  audioUpdate(input: AudioUpdateInput!): Audio!
  audioDestroy(id: ID!): Boolean! # Deletes an audio object
}

# Assuming `IDFilterType`, `FindSortType`, `Timestamp`, `Image`, `Performer`, and `Tag` types are already defined in other schema files.
```

### 2. Plugin Interaction Example (`pkg/plugin/examples/python/stash_interface.py`)

To demonstrate how plugins can interact with the new `Audio` GraphQL API, the `StashInterface` class would be extended with methods for finding, creating, and updating audio objects.

```python
# --- pkg/plugin/examples/python/stash_interface.py (Additions) ---
import requests

class StashInterface:
	port = ""
	url = ""
	headers = {
		"Accept-Encoding": "gzip, deflate, br",
		"Content-Type": "application/json",
		"Accept": "application/json",
		"Connection": "keep-alive",
		"DNT": "1"
		}

	def __init__(self, conn):
		self.port = conn['Port']
		scheme = conn['Scheme']

		self.url = scheme + "://localhost:" + str(self.port) + "/graphql"

		# Session cookie for authentication
		self.cookies = {
			'session': conn.get('SessionCookie').get('Value')
		}

	def __callGraphQL(self, query, variables = None):
		json = {}
		json['query'] = query
		if variables != None:
			json['variables'] = variables
		
		# handle cookies
		response = requests.post(self.url, json=json, headers=self.headers, cookies=self.cookies)
		
		if response.status_code == 200:
			result = response.json()
			if result.get("error", None):
				for error in result["error"]["errors"]:
					raise Exception("GraphQL error: {}".format(error))
			if result.get("data", None):
				return result.get("data")
		else:
			raise Exception("GraphQL query failed:{} - {}. Query: {}. Variables: {}".format(response.status_code, response.content, query, variables))

	# --- EXISTING METHODS (omitted for brevity) ---
	# def findTagIdWithName(self, name): ...
	# def createTagWithName(self, name): ...
	# def destroyTag(self, id): ...
	# def findRandomSceneId(self): ...
	
	# --- NEW METHODS FOR AUDIO ---

	def findAudioIdWithName(self, name):
		query = """
query findAudios($filter: AudioFilterType!) {
  findAudios(filter: $filter) {
    id
    name
  }
}
		"""
		variables = {'filter': {'q': name}}
		result = self.__callGraphQL(query, variables)
		if result and result.get("findAudios"):
			for audio in result["findAudios"]:
				if audio["name"] == name:
					return audio["id"]
		return None

	def createAudio(self, path: str, checksum: str, size: int, duration: float = None, name: str = None, author: str = None, performerIds: list[str] = None, url: str = None, website: str = None, date: str = None, language: str = None, coverFrontId: str = None, coverBackId: str = None, rating: int = None, tagIds: list[str] = None, details: str = None):
		query = """
mutation audioCreate($input: AudioCreateInput!) {
  audioCreate(input: $input) {
    id       
  }
}
		"""
		input_data = {
			'path': path,
			'checksum': checksum,
			'size': size,
		}
		if duration is not None: input_data['duration'] = duration
		if name is not None: input_data['name'] = name
		if author is not None: input_data['author'] = author
		if performerIds is not None: input_data['performerIds'] = performerIds
		if url is not None: input_data['url'] = url
		if website is not None: input_data['website'] = website
		if date is not None: input_data['date'] = date
		if language is not None: input_data['language'] = language
		if coverFrontId is not None: input_data['coverFrontId'] = coverFrontId
		if coverBackId is not None: input_data['coverBackId'] = coverBackId
		if rating is not None: input_data['rating'] = rating
		if tagIds is not None: input_data['tagIds'] = tagIds
		if details is not None: input_data['details'] = details

		variables = {'input': input_data}
		result = self.__callGraphQL(query, variables)
		return result["audioCreate"]["id"]

	def updateAudio(self, id: str, **kwargs):
		query = """
mutation audioUpdate($input: AudioUpdateInput!) {
  audioUpdate(input: $input) {
    id       
  }
}
		"""
		input_data = {'id': id}
		# Map python kwargs to GraphQL input field names (camelCase)
		field_map = {
			'performer_ids': 'performerIds',
			'tag_ids': 'tagIds',
			'cover_front_id': 'coverFrontId',
			'cover_back_id': 'coverBackId'
		}
		for key, value in kwargs.items():
			if value is not None:
				input_data[field_map.get(key, key)] = value

		variables = {'input': input_data}
		result = self.__callGraphQL(query, variables)
		return result["audioUpdate"]["id"]
    
	def destroyAudio(self, audio_id: str):
		query = """
mutation audioDestroy($id: ID!) {
  audioDestroy(id: $id)
}
		"""
		variables = {'id': audio_id}
		self.__callGraphQL(query, variables)

```

### 3. Additional Implementation Steps (Outline)

This proposed GraphQL schema and plugin interface are foundational. A complete implementation would require further work across the application:

*   **Backend (Go):**
    *   Create corresponding Go structs for the `Audio` model and its database interactions (e.g., in `pkg/models`, `pkg/database`).
    *   Implement GraphQL resolvers for `Audio` queries and mutations in `pkg/api`.
    *   Integrate audio file scanning logic (e.g., in `pkg/scanner`) to detect audio files (e.g., `.mp3`, `.flac`, `.wav`), extract metadata (ID3 tags, duration, etc.), and populate the database.
    *   Implement services for managing `Audio` objects, including relationships to `Performers`, `Tags`, and `Images`.
*   **Frontend (React/TypeScript):**
    *   Update `codegen.ts` to ensure new GraphQL types are generated into `src/core/generated-graphql.ts`. This typically works automatically if the `audio.graphql` file is placed in a scanned schema directory.
    *   Develop new UI components for listing, viewing details, and playing audio files.
    *   Implement an audio player (e.g., utilizing `video.js` or a dedicated HTML5 audio player).
    *   Create forms for editing audio metadata, including selecting performers, tags, and uploading cover art.
    *   Integrate audio into the global search and filtering capabilities.

This submission provides the architectural blueprint for integrating audio file support into Stash, focusing on the data model and API interaction points as requested by the feature.

**EVM Wallet for Payout:** 0x0c67cbE9e30c66267975eE2D74Eb88036CA65e9F
**Solana Wallet for Payout:** 3hLpMvUUS685bZz2PzR6vWeA37UrdKq924s7yLMB47eZ
**XRPL Wallet for Payout:** 0x0c67cbE9e30c66267975eE2D74Eb88036CA65e9F (EVM Sidechain)

**Comment by allornothingai** [comment]
To address the feature request for audio file/story support (Issue #1258), a new `Audio` media type needs to be introduced into the Stash application. This involves defining a new GraphQL schema, implementing corresponding backend logic for file scanning, metadata extraction, database storage, and developing a frontend UI for managing and playing audio files.

Below is a proposal for the core GraphQL schema changes and an example of how plugins would interact with the new API.

### 1. Proposed GraphQL Schema (`graphql/schema/audio.graphql`)

This new file would be added under `graphql/schema/` (e.g., `graphql/schema/audio.graphql`). This schema defines the `Audio` type and its associated input types, queries, and mutations, incorporating all the requested metadata fields.

```graphql
# Defines the Audio object
type Audio {
  id: ID!
  checksum: String! # Unique identifier for the audio file content (e.g., SHA256)
  path: String! # Full path to the audio file
  size: Int! # File size in bytes
  duration: Float # Duration of the audio in seconds

  created: Timestamp!
  updated: Timestamp!

  # User-editable metadata
  name: String
  author: String # Primary author or narrator, could be a string
  performers: [Performer!] # Array of associated performers (many-to-many relationship)
  url: String # URL to the source of the audio
  website: String # Website of the source
  date: String # Date of creation or release (e.g., YYYY-MM-DD)
  language: String # Language of the audio (e.g., "en", "es")

  cover_front: Image # Front cover image (linked via Image ID)
  cover_back: Image # Back cover image (linked via Image ID)

  rating: Int # User rating
  tags: [Tag!] # Array of associated tags (many-to-many relationship)
  details: String # Detailed description or notes
}

# Input for creating a new Audio object
input AudioCreateInput {
  path: String!
  checksum: String!
  size: Int!
  duration: Float # Optional
  
  name: String
  author: String
  performerIds: [ID!] # IDs of existing performers
  url: String
  website: String
  date: String
  language: String
  coverFrontId: ID # ID of an existing Image for the front cover
  coverBackId: ID # ID of an existing Image for the back cover
  rating: Int
  tagIds: [ID!] # IDs of existing tags
  details: String
}

# Input for updating an existing Audio object
input AudioUpdateInput {
  id: ID!
  
  name: String
  author: String
  performerIds: [ID!]
  url: String
  website: String
  date: String
  language: String
  coverFrontId: ID
  coverBackId: ID
  rating: Int
  tagIds: [ID!]
  details: String
}

# Input for filtering audio objects
input AudioFilterType {
  q: String # General search query (name, author, details, etc.)
  tags: IDFilterType # Filter by associated tags
  performers: IDFilterType # Filter by associated performers
  # Additional filters can be added here (e.g., duration range, language filter)
}

# Extend the Query type to include audio-related queries
extend type Query {
  findAudios(filter: AudioFilterType, first: Int, skip: Int, sort: FindSortType): [Audio!]!
  getAudio(id: ID!): Audio
}

# Extend the Mutation type to include audio-related mutations
extend type Mutation {
  audioCreate(input: AudioCreateInput!): Audio!
  audioUpdate(input: AudioUpdateInput!): Audio!
  audioDestroy(id: ID!): Boolean! # Deletes an audio object
}

# Assuming `IDFilterType`, `FindSortType`, `Timestamp`, `Image`, `Performer`, and `Tag` types are already defined in other schema files.
```

### 2. Plugin Interaction Example (`pkg/plugin/examples/python/stash_interface.py`)

To demonstrate how plugins can interact with the new `Audio` GraphQL API, the `StashInterface` class would be extended with methods for finding, creating, and updating audio objects.

```python
--- a/pkg/plugin/examples/python/stash_interface.py
+++ b/pkg/plugin/examples/python/stash_interface.py
@@ -62,3 +62,81 @@
 		}}
 
 		self.__callGraphQL(query, variables)
+
+	def findAudioIdWithName(self, name):
+		query = """
+query findAudios($filter: AudioFilterType!) {
+  findAudios(filter: $filter) {
+    id
+    name
+  }
+}
+		"""
+		variables = {'filter': {'q': name}}
+		result = self.__callGraphQL(query, variables)
+		if result and result.get("findAudios"):
+			for audio in result["findAudios"]:
+				if audio["name"] == name:
+					return audio["id"]
+		return None
+
+	def createAudio(self, path: str, checksum: str, size: int, duration: float = None, name: str = None, author: str = None, performerIds: list[str] = None, url: str = None, website: str = None, date: str = None, language: str = None, coverFrontId: str = None, coverBackId: str = None, rating: int = None, tagIds: list[str] = None, details: str = None):
+		query = """
+mutation audioCreate($input: AudioCreateInput!) {
+  audioCreate(input: $input) {
+    id       
+  }
+}
+		"""
+		input_data = {
+			'path': path,
+			'checksum': checksum,
+			'size': size,
+		}
+		if duration is not None: input_data['duration'] = duration
+		if name is not None: input_data['name'] = name
+		if author is not None: input_data['author'] = author
+		if performerIds is not None: input_data['performerIds'] = performerIds
+		if url is not None: input_data['url'] = url
+		if website is not None: input_data['website'] = website
+		if date is not None: input_data['date'] = date
+		if language is not None: input_data['language'] = language
+		if coverFrontId is not None: input_data['coverFrontId'] = coverFrontId
+		if coverBackId is not None: input_data['coverBackId'] = coverBackId
+		if rating is not None: input_data['rating'] = rating
+		if tagIds is not None: input_data['tagIds'] = tagIds
+		if details is not None: input_data['details'] = details
+
+		variables = {'input': input_data}
+		result = self.__callGraphQL(query, variables)
+		return result["audioCreate"]["id"]
+
+	def updateAudio(self, id: str, **kwargs):
+		query = """
+mutation audioUpdate($input: AudioUpdateInput!) {
+  audioUpdate(input: $input) {
+    id       
+  }
+}
+		"""
+		input_data = {'id': id}
+		# Map python kwargs to GraphQL input field names (camelCase)
+		field_map = {
+			'performer_ids': 'performerIds',
+			'tag_ids': 'tagIds',
+			'cover_front_id': 'coverFrontId',
+			'cover_back_id': 'coverBackId'
+		}
+		for key, value in kwargs.items():
+			if value is not None:
+				input_data[field_map.get(key, key)] = value
+
+		variables = {'input': input_data}
+		result = self.__callGraphQL(query, variables)
+		return result["audioUpdate"]["id"]
+    
+	def destroyAudio(self, audio_id: str):
+		query = """
+mutation audioDestroy($id: ID!) {
+  audioDestroy(id: $id)
+}
+		"""
+		variables = {'id': audio_id}
+		self.__callGraphQL(query, variables)
```

### 3. Additional Implementation Steps (Outline)

This proposed GraphQL schema and plugin interface are foundational. A complete implementation would require further work across the application:

*   **Backend (Go):**
    *   Create corresponding Go structs for the `Audio` model and its database interactions (e.g., in `pkg/models`, `pkg/database`).
    *   Implement GraphQL resolvers for `Audio` queries and mutations in `pkg/api`.
    *   Integrate audio file scanning logic (e.g., in `pkg/scanner`) to detect audio files (e.g., `.mp3`, `.flac`, `.wav`), extract metadata (ID3 tags, duration, etc.), and populate the database.
    *   Implement services for managing `Audio` objects, including relationships to `Performers`, `Tags`, and `Images`.
*   **Frontend (React/TypeScript):**
    *   Update `codegen.ts` to ensure new GraphQL types are generated into `src/core/generated-graphql.ts`. This typically works automatically if the `audio.graphql` file is placed in a scanned schema directory.
    *   Develop new UI components for listing, viewing details, and playing audio files.
    *   Implement an audio player (e.g., utilizing `video.js` or a dedicated HTML5 audio player).
    *   Create forms for editing audio metadata, including selecting performers, tags, and uploading cover art.
    *   Integrate audio into the global search and filtering capabilities.

This submission provides the architectural blueprint for integrating audio file support into Stash, focusing on the data model and API interaction points as requested by the feature.

**EVM Wallet for Payout:** 0x0c67cbE9e30c66267975eE2D74Eb88036CA65e9F
**Solana Wallet for Payout:** 3hLpMvUUS685bZz2PzR6vWeA37UrdKq924s7yLMB47eZ
**XRPL Wallet for Payout:** 0x0c67cbE9e30c66267975eE2D74Eb88036CA65e9F (EVM Sidechain)

**Comment by edvinPL** [comment]
btw wanted to mention that THIS:
https://discourse.stashapp.cc/t/audio-transcodes/1338
Renders audio as video (if i understand correctly) and allows playback & hoover-to-preview(listen) - though its 1). no soundwave, 2). Quite slow

@allornothingai I quite like the top-level type integration, reuse of existing relations... but I am wondering where and how you plan to handle the preview (scrubbing) of the audio (same as it's handled right now for video)?

**Comment by bob12224** [comment]
Is anyone working on this?

If not, I would be willing to work on the basic integration. I also see the great benefit of Stashapp being a media asset manager (MAM/DAM). But I will only work on it if I can get a list of mvp features that will be approved for merge. I imagine I can get a PR in for review within a week, so long as it mvp (new audio type, graphql support, alter scanner to search for audio features, use ffprobe to extract details; i won't be able to do scrubbing/thuumbnails/etc).

**Comment by Gykes** [comment]
> Is anyone working on this?

Not to my knowledge

> If not, I would be willing to work on the basic integration. I also see the great benefit of Stashapp being a media asset manager (MAM/DAM). But I will only work on it if I can get a list of mvp features that will be approved for merge. I imagine I can get a PR in for review within a week, so long as it mvp (new audio type, graphql support, alter scanner to search for audio features, use ffprobe to extract details; i won't be able to do scrubbing/thuumbnails/etc).

I'm not against having this added and helping/reviewing your work. It has been requested tons of times just no one to do the work. This would probably be a much larger refactor. I think we would need to be smart about it and not just band aid it on. UI/UX would need to be talked about. I know the top nav bar is starting to get kind of cluttered, viewing, scraping, the whole lot. The main dev did mention some other changes that would potentially make audio a little easier if we were to get them in first.  Could potentially have a chat on Discord if you have that platform

tldr; with my limited power I am okay with it and willing to help.

**Comment by bob12224** [comment]
@Gykes and any other maintainers (@WithoutPants ?), here is a copy-paste backend PR https://github.com/stashapp/stash/pull/6824
It shows the number of files for MVP, provides a good talking space (comments can be added directly to lines in files).

If we get agreement on MVP, then I will chug through this.
NOTE: I would prefer to limit the PR to backend, then followup with a frontend PR.

**Comment by Gykes** [comment]
Agreed with separate PRs. Ping me whenever you want me to take a look at it or have any specific questions.

**Comment by yw13931835525-cyber** [comment]
We are interested in this audio files/stories feature bounty.

Our team specializes in media processing and metadata extraction. We can implement full audio support with all the requested metadata fields.

Company: Hebei Gaoruan Technology Co., Ltd.
Payment: Wire transfer or crypto (USDT/ETH/BTC)

**Comment by rezreal** [comment]
This is my Band-aid, and it worked for years now... It simply allows you to have mp3s or m4as as scenes.

Add mp3,m4a as supported "video extensions"

Configure the following custom javascript
```js
PluginApi.patch.instead("ScenePlayer", function (props, params, next) {
    let scene = props.scene
    if (scene.files.some(file => file.path.endsWith('.mp3') || file.path.endsWith('.m4a'))) {
        scene = {
            ...scene,
            sceneStreams: scene.sceneStreams.filter(ss => ss.label === 'WEBM')
        }
    }
    return next({...props, scene}, params);
});
```

It will simply choose webm as encoder which seems to work for audios. phash creation (takes ages) and thumbnails obviously dont work, but you can have custom iamges on the scene which will then be displayed in the background while playback. Good enough for me. Would also be super easy to make a ui extension out of it.

**Comment by bob12224** [comment]
For those who are wanting/waiting for this. I am ready to get the development done. Just waiting on a maintainer to agree with the scope. It has been a week, I'll wait another week before abandoning it.

**Comment by Gykes** [comment]
> For those who are wanting/waiting for this. I am ready to get the development done. Just waiting on a maintainer to agree with the scope. It has been a week, I'll wait another week before abandoning it.

The main dev is taking a minor break due to real life stuff. He's still around but not actively looking at stuff like this as of this moment. 

I've already told you the overall backend is fine but have some questions about overall implementation.

If you have Discord it'll be easier there if not we can do it on your PR
